File: temp/source/pipeline/sources/extract-content.ts
import { Readability } from "@mozilla/readability";
import { parseHTML } from "linkedom";
import { log } from "../utils";
import { fetchWithTimeout } from "./utils";

// These domains can't be scraped server-side - they always return error/login pages.
const SKIP_DOMAINS: string[] = ["x.com", "twitter.com"];
const SNIPPET_MAX_LENGTH = 500;
const EXTRACT_TIMEOUT_MS = 5_000;
const CONCURRENCY = 5;

function shouldSkip(url: string): boolean {
  try {
    const host = new URL(url).hostname;
    return SKIP_DOMAINS.some((d) => host === d || host.endsWith(`.${d}`));
  } catch {
    return true;
  }
}

function ensureBody(html: string): string {
  if (/<body[\s>]/i.test(html)) return html;
  return `<html><head></head><body>${html}</body></html>`;
}

// Phrases that indicate a page blocked us instead of returning real content.
const BLOCKER_PATTERNS: RegExp[] = [
  /something went wrong.*don't fret/i,
  /disable.*privacy\s+extensions/i,
  /privacy\s+extensions.*and.*retry/i,
  /please\s+(?:sign|log)\s*in/i,
  /sign\s*in\s+to\s+(?:continue|read|access)/i,
  /log\s*in\s+to\s+(?:continue|read|access)/i,
  /subscribe\s+to\s+(?:continue|read|access|unlock)/i,
  /create\s+an?\s+account\s+to\s+(?:continue|read|access)/i,
  /enable\s+javascript\s+to\s+(?:continue|view|use)/i,
  /javascript\s+is\s+(?:disabled|required)/i,
  /access\s+denied/i,
  /403\s+forbidden/i,
];

function isBlockerContent(text: string): boolean {
  if (text.length < 50) return true; // too short to be real article text
  return BLOCKER_PATTERNS.some((re) => re.test(text));
}

function extractText(html: string, maxLength?: number): string {
  try {
    const { document } = parseHTML(ensureBody(html));
    const reader = new Readability(document as unknown as Document);
    const article = reader.parse();
    if (article) {
      const text =
        article.textContent?.replace(/\s+/g, " ").trim() ||
        article.excerpt ||
        "";
      if (!text || isBlockerContent(text)) return "";
      return maxLength !== undefined ? text.slice(0, maxLength) : text;
    }
  } catch {
    // fall through
  }
  return "";
}

async function extractHtml(url: string): Promise<string> {
  if (shouldSkip(url)) return "";

  try {
    const res = await fetchWithTimeout(url, EXTRACT_TIMEOUT_MS);
    if (!res.ok) return "";

    const contentType = res.headers.get("content-type") ?? "";
    if (!contentType.includes("text/html")) return "";

    return await res.text();
  } catch {
    return "";
  }
}

async function batchFetch(
  urls: string[],
  fn: (url: string, idx: number, total: number) => Promise<string>,
): Promise<Map<string, string>> {
  const map = new Map<string, string>();
  for (let i = 0; i < urls.length; i += CONCURRENCY) {
    const batch = urls.slice(i, i + CONCURRENCY);
    const results = await Promise.allSettled(
      batch.map((url, j) =>
        fn(url, i + j, urls.length).then((value) => ({ url, value })),
      ),
    );
    for (const r of results) {
      if (r.status === "fulfilled" && r.value.value) {
        map.set(r.value.url, r.value.value);
      }
    }
  }
  return map;
}

export async function extractContent<
  T extends { url: string; content: string },
>(articles: T[]): Promise<T[]> {
  const needsContent = articles.filter((a) => !a.content);
  if (needsContent.length === 0) return articles;

  const uniqueUrls = [...new Set(needsContent.map((a) => a.url))];
  log(`  Extracting content for ${uniqueUrls.length} URLs...`);

  const snippetMap = await batchFetch(uniqueUrls, async (url) =>
    extractText(await extractHtml(url), SNIPPET_MAX_LENGTH),
  );

  log(`  Extracted ${snippetMap.size}/${uniqueUrls.length} snippets`);

  return articles.map((a) => {
    const snippet = snippetMap.get(a.url);
    return snippet ? { ...a, content: snippet } : a;
  });
}

export async function reExtractFullContent(
  articles: { url: string }[],
): Promise<Map<string, string>> {
  if (articles.length === 0) return new Map();

  const uniqueUrls = [...new Set(articles.map((a) => a.url))];
  log(`  Re-extracting full content for ${uniqueUrls.length} URLs...`);

  const contentMap = await batchFetch(uniqueUrls, async (url, idx, total) => {
    log(`    [${idx + 1}/${total}] Fetching: ${url}`);
    const t0 = Date.now();
    const content = extractText(await extractHtml(url));
    const ms = Date.now() - t0;
    log(
      `    [${idx + 1}/${total}] ${content ? `OK (${content.length} chars, ${ms}ms)` : `no content (${ms}ms)`}: ${url}`,
    );
    return content;
  });

  log(`  Re-extracted ${contentMap.size}/${uniqueUrls.length} full texts`);
  return contentMap;
}

File: backend/pipeline/sources/extract_content.py
from __future__ import annotations

import re
from concurrent.futures import ThreadPoolExecutor, as_completed

import trafilatura

from pipeline.utils import log
from pipeline.sources.utils import fetch_with_timeout

SKIP_DOMAINS: set[str] = {"x.com", "twitter.com"}
SNIPPET_MAX_LENGTH = 500
EXTRACT_TIMEOUT_SECS = 5.0
CONCURRENCY = 5

BLOCKER_PATTERNS: list[re.Pattern] = [
    re.compile(r"something went wrong.*don't fret", re.IGNORECASE | re.DOTALL),
    re.compile(r"disable.*privacy\s+extensions", re.IGNORECASE),
    re.compile(r"privacy\s+extensions.*and.*retry", re.IGNORECASE),
    re.compile(r"please\s+(?:sign|log)\s*in", re.IGNORECASE),
    re.compile(r"sign\s*in\s+to\s+(?:continue|read|access)", re.IGNORECASE),
    re.compile(r"log\s*in\s+to\s+(?:continue|read|access)", re.IGNORECASE),
    re.compile(r"subscribe\s+to\s+(?:continue|read|access|unlock)", re.IGNORECASE),
    re.compile(r"create\s+an?\s+account\s+to\s+(?:continue|read|access)", re.IGNORECASE),
    re.compile(r"enable\s+javascript\s+to\s+(?:continue|view|use)", re.IGNORECASE),
    re.compile(r"javascript\s+is\s+(?:disabled|required)", re.IGNORECASE),
    re.compile(r"access\s+denied", re.IGNORECASE),
    re.compile(r"403\s+forbidden", re.IGNORECASE),
]


def _should_skip(url: str) -> bool:
    try:
        from urllib.parse import urlparse
        host = urlparse(url).hostname or ""
        return any(host == d or host.endswith(f".{d}") for d in SKIP_DOMAINS)
    except Exception:
        return True


def _is_blocker(text: str) -> bool:
    if len(text) < 50:
        return True
    return any(p.search(text) for p in BLOCKER_PATTERNS)


def _fetch_html(url: str) -> str:
    if _should_skip(url):
        return ""
    try:
        resp = fetch_with_timeout(url, timeout=EXTRACT_TIMEOUT_SECS)
        if not resp.is_success:
            return ""
        ct = resp.headers.get("content-type", "")
        if "text/html" not in ct:
            return ""
        return resp.text
    except Exception:
        return ""


def _extract_text(html: str, max_length: int | None = None) -> str:
    if not html:
        return ""
    text = trafilatura.extract(html, include_comments=False, include_tables=False) or ""
    text = re.sub(r"\s+", " ", text).strip()
    if not text or _is_blocker(text):
        return ""
    return text[:max_length] if max_length else text


def _extract_one_snippet(url: str) -> tuple[str, str]:
    return url, _extract_text(_fetch_html(url), SNIPPET_MAX_LENGTH)


def _extract_one_full(url: str, idx: int, total: int) -> tuple[str, str]:
    log(f"    [{idx + 1}/{total}] Fetching: {url}")
    html = _fetch_html(url)
    text = _extract_text(html)
    log(f"    [{idx + 1}/{total}] {'OK (' + str(len(text)) + ' chars)' if text else 'no content'}: {url}")
    return url, text


def extract_content(articles: list[dict]) -> list[dict]:
    """Fill in missing content snippets for a list of article dicts."""
    needs = [a for a in articles if not a.get("content")]
    if not needs:
        return articles

    unique_urls = list(dict.fromkeys(a["url"] for a in needs))
    log(f"  Extracting content for {len(unique_urls)} URLs...")

    snippet_map: dict[str, str] = {}
    with ThreadPoolExecutor(max_workers=CONCURRENCY) as executor:
        futures = {executor.submit(_extract_one_snippet, url): url for url in unique_urls}
        for future in as_completed(futures):
            try:
                url, text = future.result()
                if text:
                    snippet_map[url] = text
            except Exception:
                pass

    log(f"  Extracted {len(snippet_map)}/{len(unique_urls)} snippets")

    result = []
    for a in articles:
        if not a.get("content") and a["url"] in snippet_map:
            result.append({**a, "content": snippet_map[a["url"]]})
        else:
            result.append(a)
    return result


def re_extract_full_content(articles: list[dict]) -> dict[str, str]:
    """Re-fetch full article text for already-inserted articles. Returns url->text map."""
    if not articles:
        return {}

    unique_urls = list(dict.fromkeys(a["url"] for a in articles))
    log(f"  Re-extracting full content for {len(unique_urls)} URLs...")

    content_map: dict[str, str] = {}
    total = len(unique_urls)

    with ThreadPoolExecutor(max_workers=CONCURRENCY) as executor:
        futures = {
            executor.submit(_extract_one_full, url, idx, total): url
            for idx, url in enumerate(unique_urls)
        }
        for future in as_completed(futures):
            try:
                url, text = future.result()
                if text:
                    content_map[url] = text
            except Exception:
                pass

    log(f"  Re-extracted {len(content_map)}/{len(unique_urls)} full texts")
    return content_map
